### PR TITLE
tests and changes for __cnfLine

### PR DESCRIPTION
**Pull request by t0mpr1c3**
Referencing https://github.com/AllYarnsAreBeautiful/ayab-desktop/issues/301, https://github.com/AllYarnsAreBeautiful/ayab-desktop/pull/322#issuecomment-643115681

### CODE REVIEW

File: src/main/python/ayab/plugins/ayab_plugin/ayab_control.py [hunk 1]
```diff
@@ -370,82 +376,95 @@ def __middlecoltwice(self, lineNumber, imgHeight, lenImgExpanded):
...
+        imgRow = self._startLine + q
 
-        if self.__infRepeat:
+        if self._infRepeat:
             imgRow = imgRow % imgHeight
 
-        if imgRow % 2 != 0:
-            color = ((lineNumber % passesPerRow) + 1) // 2
+        if firstCol or lastCol:
+            color = (lastCol + q) % 2
```
Comment (t0mpr1c3) [inline src/main/python/ayab/plugins/ayab_plugin/ayab_control.py:391]: This changes corrects what I believe to be a bug that would occur with infinite repeats and image files with an odd number of rows.

File: src/main/python/ayab/plugins/ayab_plugin/ayab_control.py [hunk 2]
```diff
@@ -85,20 +91,20 @@ def get_progress(self):
...
-            self.__knit_func = self.__circular_ribber
+        elif self.__knitting_mode == KnittingMode.CLASSIC_RIBBER_1.value and self._numColors == 2:
+            self.__knit_func = self._doublebed_2col
+        elif self.__knitting_mode == KnittingMode.CLASSIC_RIBBER_1.value and self._numColors > 2:
+            self.__knit_func = self._doublebed_multicol
+        elif self.__knitting_mode == KnittingMode.MIDDLECOLORSTWICE_RIBBER.value and self._numColors >= 2:
+            self.__knit_func = self._middlecoltwice
+        elif self.__knitting_mode == KnittingMode.HEARTOFPLUTO_RIBBER.value and self._numColors >= 2:
+            self.__knit_func = self._heartofpluto
+        elif self.__knitting_mode == KnittingMode.CIRCULAR_RIBBER.value: # not restricted to 2 colors
```
Comment (t0mpr1c3) [inline src/main/python/ayab/plugins/ayab_plugin/ayab_control.py:104]: I could not see any necessary reason to restrict circular knitting to 2 colours. Am I missing something?

File: src/main/python/ayab/plugins/ayab_plugin/ayab_control.py [hunk 3]
```diff
@@ -370,82 +376,95 @@ def __middlecoltwice(self, lineNumber, imgHeight, lenImgExpanded):
...
+    def _heartofpluto(self, lineNumber, imgHeight, lenImgExpanded):
 
-        # Double the line minus the 2 you save from
-        # early advancing to next row
-        passesPerRow = 2 * self.__numColors - 2
+        # doublebed <3 of pluto multicolor
+        # 0-00 1-11 2-22 3-33 4-44 5-55 .. (imgRow)
+        # 0123 4567 8911 1111 1111 2222.. (lineNumber)
+        #             01 2345 6789 0123
+        #
```
Comment (t0mpr1c3) [inline src/main/python/ayab/plugins/ayab_plugin/ayab_control.py:411]: I added this multiline comment to help show what I believe the Heart of Pluto mode should be doing. Is it correct?

File: src/main/python/ayab/plugins/ayab_plugin/ayab_control.py [hunk 4]
```diff
@@ -370,82 +376,95 @@ def __middlecoltwice(self, lineNumber, imgHeight, lenImgExpanded):
...
+        lastCol = (r == passesPerRow - 1)
 
-        indexToSend = imgRow * self.__numColors
+        imgRow = self._startLine + q
 
-        # check if it's time to send a blank line
-        if lineNumber % passesPerRow != 0 and lineNumber % 2 == 0:
-            color = 0
-            sendBlankLine = True
-        else:
```
Comment (t0mpr1c3) [inline src/main/python/ayab/plugins/ayab_plugin/ayab_control.py:412]: This clause contains errors introduced in the previous commit when code was removed from __cnfLine()

File: src/main/python/ayab/plugins/ayab_plugin/ayab_control.py [hunk 5]
```diff
@@ -370,82 +376,95 @@ def __middlecoltwice(self, lineNumber, imgHeight, lenImgExpanded):
...
-        sendBlankLine = (lineNumber % 2 == 1)
+        if self._infRepeat:
+            h  = h % lenImgExpanded
 
-        lastLine = (indexToSend == lenImgExpanded - 1) and (sendBlankLine is True)
+        q, color = divmod(h, self._numColors)
+
+        # not strictly necessary to take modulus
+        imgRow = (self._startLine + q) % imgHeight
+
```
Comment (t0mpr1c3) [inline src/main/python/ayab/plugins/ayab_plugin/ayab_control.py:461]: Code generalized for > 2 colors

File: src/main/python/ayab/plugins/ayab_plugin/ayab_control.py (hunk 2 above)
Comment (chris007de) [reply src/main/python/ayab/plugins/ayab_plugin/ayab_control.py:104]: I have no idea ;) @Adrienne200 might be able to answer that question. Meanwhile, perhaps https://manual.ayab-knitting.com/how_to_knit/circular_knitting/ gives an answer.

**Comment by chris007de** [comment]
:tada: Very nice, such a big improvement over the legacy code!
Perhaps we also can add some tests for startLine != 0 in the various knitting modes, as we sometimes had trouble with that.

**Pull request merged**